src/worker/builder.rs

```diff
@@ -90,6 +96,65 @@ impl<E: 'static> WorkerBuilder<E> {
         self
     }
 
+    /// Set a graceful shutdown signal.
+    ///
+    /// As soon as the provided future resolves, the graceful shutdown will step in
+    /// making the long-running operation (see [`Worker::run`]) return control to the calling code.
+    ///
+    /// The graceful shutdown itself is a race between the clean up needed to be performed
+    /// (e.g. report on the currently processed to the Faktory server) and a shutdown deadline.
+    /// The latter can be customized via [`WorkerBuilder::graceful_shutdown_period`].
+    ///
+    /// ```no_run
+    /// # tokio_test::block_on(async {
+    /// use faktory::{Client, Job, Worker};
+    /// use tokio_util::sync::CancellationToken;
+    ///
+    /// Client::connect(None)
+    ///     .await
+    ///     .unwrap()
+    ///     .enqueue(Job::new("foobar", vec!["z"]))
+    ///     .await
+    ///     .unwrap();
+    ///
+    /// // create a signalling future (we are using a utility from the `tokio_util` crate)
+    /// let token = CancellationToken::new();
+    /// let child_token = token.child_token();
+    /// let signal = async move { child_token.cancelled().await };
+    ///
+    /// // get a connected worker
+    /// let mut w = Worker::builder()
+    ///     .with_graceful_shutdown(signal)
+    ///     .register_fn("job_type", move |_| async { Ok::<(), std::io::Error>(()) })
+    ///     .connect(None)
+    ///     .await
+    ///     .unwrap();
+    ///
+    /// // start consuming
+    /// let jh = tokio::spawn(async move { w.run(&["default"]).await });
+    ///
+    /// // send a signal to eventually return control (upon graceful shutdown)
+    /// token.cancel();
+    /// # });
+    /// ```
+    pub fn with_graceful_shutdown<F>(mut self, signal: F) -> Self
+    where
+        F: Future<Output = ()> + 'static + Send,
+    {
+        self.shutdown_signal = Some(Box::pin(signal));
+        self
+    }
+
+    /// Set the graceful shutdown period in milliseconds. Defaults to 5000.
+    ///
+    /// This will be used once the worker is sent a termination signal whether it is at the application
+    /// (via a signalling future, see [`WorkerBuilder::with_graceful_shutdown`]) or OS level (via Ctrl-C signal,
+    /// see [`Worker::run_to_completion`]).
+    pub fn graceful_shutdown_period(mut self, millis: u64) -> Self {
```

Comment:
This should take a `Duration`.

Also, should we allow users to _not_ set a shutdown period (i.e., to never have shutdown aborted)?

Reply:
Updated.

My understanding is when dealing with futures or long-running tasks in general you almost always want to have a timeout. It's just that many (most?) of the operations we usually deal with are quick enough to finish for us to not get into a "hanging" state.

At the same time the question is where did I get the default value from (and the answer may not sound convincing). I also just looked up the `reqwest` API and they are defaulting the [timeout](https://docs.rs/reqwest/latest/reqwest/struct.ClientBuilder.html#method.timeout) to `None` .

Yes, I think it should be defaulted to `None` and should be called simply `shutdown_timeout` (even if it's being performed with disgrace :smile: ) following the `reqwest`'s namings.